### PR TITLE
LUCENE-9508: Fix DocumentsWriter to block threads until unstalled

### PR DESCRIPTION
**Pull request by s1monw**
DWStallControl expects the caller to loop on top of the wait call to make
progress with flushing if the DW is stalled. This logic wasn't applied such that
DW only stalled for one second and then released the indexing thread. This can cause
OOM if for instance during a full flush one DWPT gets stuck and other threads keep on
indexing.

### CODE REVIEW

**Comment by s1monw** [comment]
@mikemccand can you take a look pls

File: lucene/core/src/test/org/apache/lucene/index/TestIndexWriter.java [hunk 1]
```diff
@@ -4258,4 +4258,47 @@ public void testPendingNumDocs() throws Exception {
       }
     }
   }
+
+  public void testIndexWriterBlocksOnStall() throws IOException, InterruptedException {
+    try (Directory dir = newDirectory()) {
+      try (IndexWriter writer = new IndexWriter(dir, newIndexWriterConfig())) {
```
Comment (dweiss) [inline lucene/core/src/test/org/apache/lucene/index/TestIndexWriter.java:4264]: pull it up into the same try (semicolon-separated)?

File: lucene/core/src/java/org/apache/lucene/index/DocumentsWriter.java [hunk 2]
```diff
@@ -371,19 +371,15 @@ public void close() throws IOException {
   private boolean preUpdate() throws IOException {
     ensureOpen();
     boolean hasEvents = false;
-
-    if (flushControl.anyStalledThreads() || (flushControl.numQueuedFlushes() > 0 && config.checkPendingFlushOnUpdate)) {
```
Comment (dweiss) [inline lucene/core/src/java/org/apache/lucene/index/DocumentsWriter.java:375]: oh, wonderful.

**Review by dweiss — APPROVED** [review]
LGTM.

File: lucene/core/src/test/org/apache/lucene/index/TestIndexWriter.java (hunk 1 above)
Comment (s1monw) [reply lucene/core/src/test/org/apache/lucene/index/TestIndexWriter.java:4264]: it's more clear that way what is closed first which is important here. That's why I do it this way.

File: lucene/core/src/java/org/apache/lucene/index/DocumentsWriter.java (hunk 2 above)
Comment (s1monw) [reply lucene/core/src/java/org/apache/lucene/index/DocumentsWriter.java:375]: yeah, details...

**Review by dnhatn — APPROVED** [review]
Great catch! LGTM.

**Pull request merged**